h2/src/main/org/h2/value/Value.java

```diff
@@ -882,8 +882,13 @@ private ValueByte convertToByte(Object column) {
         case REAL:
         case DOUBLE:
             return ValueByte.get(convertToByte(convertToLong(getDouble(), column), column));
-        case VARBINARY:
-            return ValueByte.get((byte) Integer.parseInt(getString(), 16));
+        case VARBINARY: {
+            byte[] bytes = getBytesNoCopy();
+            if (bytes.length == 1) {
```

Comment:
surely the logic here and below should be <= rather than == ?
i.e. if the number of bytes is <= than the destination scalar type, then convert the  bytes by zero extension?

Reply:
I have no idea why H2 allows conversion between numeric and binary string data types, such conversion has a very low meaning. The current implementation here only ensures that conversion from `TINYINT` to `VARBINARY` and then back to `TINYINT` will return the same `TINYINT` value. Old implementation was insane: some overflows were silently truncated, some were not with a weird exception.

If you need a more advanced logic here, please, describe it in more clear way with all corner cases.

Reply:
Then we should rather just remove the conversion logic for VARBINARY->all_integer_types.
I also cannot imagine that this is being used anywhere useful